src/object/srv_obj.c

```diff
@@ -854,75 +978,40 @@ ds_obj_rw_handler(crt_rpc_t *rpc)
 		D_GOTO(out, rc = -DER_STALE);
 	}
 
-	if (orw->orw_map_ver < map_ver) {
-		D_DEBUG(DB_IO, "stale version req %d map_version %d\n",
-			orw->orw_map_ver, map_ver);
-		if (update)
-			D_GOTO(out, rc = -DER_STALE);
-
-		/* It is harmless if fetch with old pool map version. */
-	} else if (orw->orw_map_ver > map_ver) {
-		/* XXX: Client (or leader replica) has newer pool map than
-		 *	current replica. Two possibile cases:
-		 *
-		 *	1. The current replica was the old leader if with
-		 *	   the old pool map version. According to current
-		 *	   leader election algorithm, it is still the new
-		 *	   leader with the new pool map version. Since no
-		 *	   leader switch, the unmatched pool version will
-		 *	   not affect DTX related availability check.
-		 *
-		 *	2. The current replica was NOT the old leader if
-		 *	   with the old pool map version. But it becomes
-		 *	   the new leader with the new pool map version.
-		 *	   In the subsequent modificaiton, it may hit
-		 *	   some 'prepared' DTX when make availability
-		 *	   check, it will return -DER_INPROGRESS that
-		 *	   will cause client to retry. It is possible
-		 *	   that the pool map version event arrives at
-		 *	   this server during the client retry. It is
-		 *	   inefficient, but harmless.
-		 */
-	}
-
-	if (!update) {
-		D_TIME_END(tls->ot_sp, OBJ_PF_UPDATE_PREP);
+	/* Single replica update or fetch. */
+	if (orw->orw_shard_tgts.ca_arrays == NULL) {
+		/* No need resend check? XXX */
 		D_TIME_START(tls->ot_sp, OBJ_PF_UPDATE_LOCAL);
-
-		rc = ds_obj_rw_local_hdlr(rpc, tag, cont_hdl, cont,
-					  &ioh, false, map_ver, dth);
-		if (rc != 0)
-			D_ERROR(DF_UOID": rw_local_hdlr (fetch) failed %d.\n",
+		rc = ds_obj_rw_local_hdlr(rpc, cont_hdl, cont, dth);
+		if (rc != 0) {
+			D_ERROR(DF_UOID": rw_local_hdlr (update) failed %d.\n",
 				DP_UOID(orw->orw_oid), rc);
+			D_GOTO(out, rc);
+		}
 
 		D_TIME_END(tls->ot_sp, OBJ_PF_UPDATE_LOCAL);
-
 		D_GOTO(out, rc);
 	}
 
-	if (!daos_oc_echo_type(daos_obj_id2class(orw->orw_oid.id_pub))) {
-		/* Check DTX configuration. */
-		rc = ds_obj_check_dtx_config(orw->orw_flags, &orw->orw_dti);
-		if (rc != 0)
+	/* Handle resend. */
+	if (orw->orw_flags & ORF_RESEND) {
+		rc = dtx_handle_resend(cont->sc_hdl, &orw->orw_oid,
+				       &orw->orw_dti,
+				       orw->orw_dkey_hash, false);
+		if (rc == -DER_ALREADY) {
+			D_GOTO(out, rc = 0);
+		} if (rc == 0) {
```

Comment:
(style)  trailing statements should be on next line (or did you mean 'else if'?)